### DISCUSSION TITLE
Q: 能不能URL规则点 || Q: Can URL rules point

### DISCUSSION THREAD

**Discussion by hulang** [walinejs/waline#1963] [General]
### Describe the bug | 问题描述

https://waline.js.org/reference/server/api.html
能不能URL规则点：
GET /api/comment/list  列表
POST /api/comment/insert  添加
POST /api/comment/update  编辑
POST /api/comment/like  点赞
POST /api/comment/status  通过/审核/垃圾
POST /api/comment/sticky  置顶
怎么实现这样，后端代码总是在一个comment方法里面写判断，代码写的稀烂，我想在comment控制器里面，定义各种方法。。这样结构清晰。

### Website URL | 问题网站

https://waline.js.org/reference/server/api.html

### Where your waline deploy? | 服务部署在哪里？

Vercel (Default)

### Where your comment data store? | 数据存储在哪里？

LeanCloud(https://leancloud.app)

---
### Describe the bug | Problem description

https://waline.js.org/reference/server/api.html
Can you point to URL rules:
GET /api/comment/list list
POST /api/comment/insert add
POST /api/comment/updateEdit
POST /api/comment/like like
POST /api/comment/status pass/review/trash
POST /api/comment/sticky sticky
How to achieve this, the back-end code always writes judgments in a comment method, the code is poorly written, I want to define various methods in the comment controller. . This structure is clear.

### Website URL | Problem Website

https://waline.js.org/reference/server/api.html

### Where your waline deploy? | Where is the service deployed?

Vercel (Default)

### Where your comment data store? | Where is your comment data store?

LeanCloud (https://leancloud.app)

**Comment by lizheming** [comment]
Waline API 是基于 RESTful API 语义的，除了 `GET /api/comment` 评论列表那块因为耦合了最近评论和前台评论列表以及后台评论列表的操作，用 type 区分没有做很好的拆分有点恶心之外，我觉得 `POST /api/comment` 和 `PUT /api/comment` 的实现还是很好的，没什么问题的。

基于功能语义的 API 确实能让每个功能更好的拆分，但它不好的地方在于会写非常多的重复代码。比如你列举的这几个 API，其实本质就是去修改 comment 表的某个字段，但如果拆分4个控制器的话，这个同样的数据库操作你就要写四遍。

> POST /api/comment/update 编辑
> POST /api/comment/like 点赞
> POST /api/comment/status 通过/审核/垃圾
> POST /api/comment/sticky 置顶

Waline API 的 RESTful 语义不算特别完美，但也不至于稀烂~ 你可以先说下你要做什么以及你的诉求，我看下怎么组织处理会比较好？

---
Waline API is based on RESTful API semantics, except for `GET /api/comment` comment list, because it is coupled with recent comments, foreground comment list and background comment list operations, it is a bit disgusting to use type to distinguish without doing a good split In addition, I think the implementation of `POST /api/comment` and `PUT /api/comment` is still very good, there is no problem.

The API based on functional semantics can indeed make each function better split, but its disadvantage is that it will write a lot of repetitive code. For example, the APIs you listed are actually to modify a certain field in the comment table, but if you split the four controllers, you have to write the same database operation four times.

> POST /api/comment/update edit
> POST /api/comment/like Like
> POST /api/comment/status pass/review/trash
> POST /api/comment/sticky sticky

The RESTful semantics of the Waline API are not particularly perfect, but they are not rotten~ You can first talk about what you want to do and your demands, and let me see how to organize and handle it better?

**Reply by hulang** [reply]
POST /api/comment 和 PUT /api/comment 的实现还是很好的，没什么问题的。你还少说了删除：Delete。这3个可以。
没有，我是习惯了。一个控制器拆分开很多对应的方法来。。目前就是在一个方法里面写。。总是感觉很稀碎。。
比如我的用的是`comment`控制器。
```php
<?php
class Comment
{
    /**
     * 加载列表
     *
     * @return mixed
     */
    public function list()
    {
        // json
        $json = [];
        return json($json);
    }
    /**
     * 添加
     *
     * @return mixed
     */
    public function insert()
    {
        // json
        $json = [];
        return json($json);
    }
    /**
     * 编辑
     *
     * @return mixed
     */
    public function update()
    {
        // json
        $json = [];
        return json($json);
    }
    /**
     * 点赞
     *
     * @return mixed
     */
    public function like()
    {
        // json
        $json = [];
        return json($json);
    }
}
```

---
The implementation of POST /api/comment and PUT /api/comment is still very good, there is no problem. You also said less about delete: Delete. These 3 are ok.
No, I'm used to it. A controller is split into many corresponding methods. . Currently it is written in a method. . It always feels shattered. .
For example, I use the `comment` controller.
```php
<?php
class comment
{
    /**
     * load list
     *
     * @return mixed
     */
    public function list()
    {
        // json
        $json = [];
        return json($json);
    }
    /**
     * Add to
     *
     * @return mixed
     */
    public function insert()
    {
        // json
        $json = [];
        return json($json);
    }
    /**
     * edit
     *
     * @return mixed
     */
    public function update()
    {
        // json
        $json = [];
        return json($json);
    }
    /**
     * like
     *
     * @return mixed
     */
    public function like()
    {
        // json
        $json = [];
        return json($json);
    }
}
```

**Reply by lizheming** [reply]
嗯嗯，了解了。感谢你的反馈，`GET /api/comment` 那个我也挺头疼，初版着急写，也没想到很好的语义化表达，就先这样了。这个我最近优化一下，做下逻辑拆分应该就可以了。

---
Hmm, got it. Thank you for your feedback. `GET /api/comment` is also a headache for me. I was in a hurry to write the first version, and I didn’t expect a good semantic expression, so let’s do it for now. I have optimized this recently, and it should be fine to do a logical split.

**Reply by hulang** [reply]
对了。我想问下。我指定的serverURL：https://域名/addons/waline/api/
删除后
请求 URL：https://域名/addons/waline/api/comment/2?lang=zh-CN
请求方法: DELETE
我不知道该怎么获取2了，不进入comment方法了？
我只会：https://域名/addons/waline/api/comment?objectId=2&lang=zh-CN
通过/待审核/垃圾/置顶/编辑/删除/点赞  都是：https://域名/addons/waline/api/comment/2?lang=zh-CN
只是请求方法不一样。。。

---
correct. I want to ask. The serverURL I specified: https://domain name/addons/waline/api/
after deletion
Request URL: https://domain name/addons/waline/api/comment/2?lang=zh-CN
Request method: DELETE
I don't know how to get 2, don't enter the comment method?
I only know: https://domain name/addons/waline/api/comment?objectId=2&lang=zh-CN
Pass/Pending review/Trash/Top/Edit/Delete/Like: https://domain name/addons/waline/api/comment/2?lang=zh-CN
Just the request method is different. . .

**Reply by Mister-Hope** [reply]
你随便改api，目前client也没有定案。我在写预设的时候遇到了问题，打算直接把markdownRenderer拆出来。

---
You can change the api at will, and the client has not yet finalized. I encountered a problem when I wrote the preset, and I planned to remove the markdownRenderer directly.

**Reply by hulang** [reply]
> 你随便改api，目前client也没有定案。我在写预设的时候遇到了问题，打算直接把markdownRenderer拆出来。
> 
> You can change the api at will, and the client has not yet finalized. I encountered a problem when I wrote the preset, and I planned to remove the markdownRenderer directly.

怎么改API啊。。。

**Reply by Mister-Hope** [reply]
我在回复 @lizheming 

---
I'm replying to @lizheming

**Reply by lizheming** [reply]
@hulang 读了几遍你的话，结合了下你的个人项目语言，我猜测你的需求应该是想复用 Waline 客户端，使用 PHP 重新实现一遍 Server API 吧。如果你使用了 PHP 的框架的话，可以查下对应框架的文档怎么获取 RESTful URL 中的资源 id。如果是纯使用 PHP 的话是需要自己解析一下的，这里提供一个简单的例子：

```php
<?php
// GET /api/comment/2
$uri = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
$uri = explode( '/', $uri );
$id = $uri[3];

```

最后，还是建议先简单了解下 RESTful，对你的实现是有帮助的



---
@hulang After reading your words several times, combined with your personal project language, I guess your requirement should be to reuse the Waline client, and use PHP to re-implement the Server API again. If you use the PHP framework, you can check the documentation of the corresponding framework how to obtain the resource id in the RESTful URL. If you use PHP purely, you need to analyze it yourself. Here is a simple example:

```php
<?php
// GET /api/comment/2
$uri = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
$uri = explode( '/', $uri );
$id = $uri[3];

```

Finally, it is recommended to have a brief understanding of RESTful first, which is helpful for your implementation

**Reply by hulang** [reply]
你这样弄。就更加复杂了，根本进入不了comment方法。。我正在尝试改：api/comment.ts文件。进行编译。
/comment/2?lang=zh-CN
改成
/comment?objectId=2&lang=zh-CN

---
You do this. It's even more complicated, and you can't enter the comment method at all. . I'm trying to modify: api/comment.ts file. to compile.
/comment/2?lang=zh-CN
changed to
/comment?objectId=2&lang=zh-CN

**Reply by lizheming** [reply]
一般 RESTful 是要做路由映射的，自己多了解一下哈。PHP 很多框架都自带 RESTful 的支持的，能省很多事情。比如这是 ThinkPHP 关于 RESTful 路由的文档 https://www.kancloud.cn/manual/thinkphp6_0/1037501

不过如果你确实没办法理解 RESTful，或者觉得它比较复杂的话，改客户端也是可以的。


---
Generally, RESTful needs to do routing mapping, so learn more about it yourself. Many PHP frameworks come with RESTful support, which can save a lot of work. For example, this is ThinkPHP's documentation on RESTful routing https://www.kancloud.cn/manual/thinkphp6_0/1037501

But if you really can't understand RESTful, or think it is more complicated, you can change the client.

**Reply by hulang** [reply]
我明白了你这个，原来是路由。。你一直说：RESTful URL，是我不理解，还是你不理解RESTful URL？我咋调用阿里云，百度的，都只有URL，参数都是post。
列表可以GET，POST都是带数据了，不会有：2?lang=zh-CN，这两个都在post里面。。

---
I understand you, it turned out to be routing. . You keep saying: RESTful URL, I don't understand, or you don't understand RESTful URL? Why do I call Alibaba Cloud and Baidu, there are only URLs, and the parameters are all post.
The list can be GET and POST with data, there will be no: 2?lang=zh-CN, both of which are in the post. .

**Reply by Mister-Hope** [reply]
I agree with hulang's opinion that leaving lang as url params with methods that have body (e.g.: post) could be confusing.

We can just redesign the whole api system while adding these new api in v1.

BTW, the `@v2` version range in CDN is there in docs since v2 initial release, so we can expect no one will accidentally upgrade to v3  client once it touches stable.

**Reply by lizheming** [reply]
@hulang 你这一系列问题我看不懂，我只能回答第一个问题，是你不理解 RESTful URL

---
@hulang I can't understand your series of questions, I can only answer the first question, you don't understand RESTful URL

**Reply by lizheming** [reply]
@Mister-Hope 我没有理解你的问题，我认为 POST 请求有 GET 参数是一个很正常的事情。

---
@Mister-Hope I didn't understand your question, I think it's a very normal thing for POST request to have GET parameters.

**Reply by hulang** [reply]
> @Mister-Hope 我没有理解你的问题，我认为 POST 请求有 GET 参数是一个很正常的事情。
> 
> @Mister-Hope I didn't understand your question, I think it's a very normal thing for POST request to have GET parameters.

POST 请求有 GET 参数是一个很正常的事情，我笑了哈，，从来没有见过阿里云，百度，他们提供API接口的，还有POST里面带GET的。。

**Reply by lizheming** [reply]
@hulang 你没见过并不一定是错误的，~~我没有时间去找一个例子给你去验证这个事情，后续找到的话我给你发一下~~ 在 GitHub 和你一直在提的百度随便找了两个 API 给你你可以一下，既有 Body 参数又有 URL 参数的。另外，我已经是非常客气的在和你沟通了，在我看来你确实不知道的东西有很多，我也并没有高高在上的和你沟通。如果你一直是这种阴阳怪气的语气沟通的话，后续的评论我就不回了。Waline 项目本身反正也是开源的，你可以按照你的需要随便改。

- https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#create-or-update-file-contents
- https://cloud.baidu.com/doc/BOS/s/zkc5ne644



---
@hulang It’s not necessarily wrong that you haven’t seen it. ~~ I don’t have time to find an example for you to verify this matter. If I find it later, I will send it to you~~ Find it on GitHub and Baidu you have been mentioning There are two APIs for you, you can look at it, there are both Body parameters and URL parameters. In addition, I have communicated with you very politely. In my opinion, there are many things that you really don't know, and I have not communicated with you in a superior manner. If you have always communicated in this yin and yang tone, I will not reply to the follow-up comments. The Waline project itself is open source anyway, and you can change it to suit your needs.

- https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#create-or-update-file-contents
- https://cloud.baidu.com/doc/BOS/s/zkc5ne644
